### PR TITLE
Provide _original_response in AppEngine adapter

### PR DESCRIPTION
**Pull request by erlichmen**
_original_response is missing from the AppEngine adapter this breaks many flows like cookie parsing. This is an attempt to fix this issue.

### CODE REVIEW

**Comment by Lukasa** [comment]
@jonparrott Can I get your eyes on this patch?

**Comment by jonparrott** [comment]
This seems reasonable.

File: urllib3/response.py [hunk 1]
```diff
@@ -132,6 +132,7 @@ def __init__(self, body='', headers=None, status=0, version=0, reason=None,
         self._fp = None
         self._original_response = original_response
         self._fp_bytes_read = 0
+        self._msg = msg
```
Comment (sigmavirus24) [inline urllib3/response.py]: Is there a reason we need to use a property for this?

File: urllib3/response.py [hunk 2]
```diff
@@ -191,6 +192,13 @@ def data(self):
     def connection(self):
         return self._connection
 
+    def isclosed(self):
+        return is_fp_closed(self._fp)
+
+    @property
+    def msg(self):
```
Comment (sigmavirus24) [inline urllib3/response.py]: I'd prefer we add a docstring (or comment) that explains why we added this.

File: urllib3/response.py (hunk 1 above)
Comment (erlichmen) [reply urllib3/response.py]: msg has the headers in list from (headers has it in dict).
This is important for the Set-Cookie and  Set-Cookie2 headers that can are multiple-values-same-key headers that the headers by combining them into a single key lose the ability to parse them and set the value accordingly in the session header.

msg is also used in the _original_response context only by the stack which expects Response object of httplib2. A different solution might be to subclass this class and put msg only in the sub class and make _original_response instantiate that new sub class but this looks like a bit overrkill for the problem in hand.

File: urllib3/contrib/appengine.py [hunk 3]
```diff
@@ -236,13 +236,21 @@ def _urlfetch_response_to_http_response(self, urlfetch_resp, **response_kw):
...
             urlfetch_resp.headers['transfer-encoding'] = ','.join(encodings)
 
+        original_response = HTTPResponse(
+          # In order for decoding to work, we must present the content as
+          # a file-like object.
+          body=BytesIO(urlfetch_resp.content),
+          msg=urlfetch_resp.header_msg,
+          headers=urlfetch_resp.headers,
+          status=urlfetch_resp.status_code,
+          **response_kw)
```
Comment (Lukasa) [inline urllib3/contrib/appengine.py]: The indentation here is off: can you indent another two spaces?

File: urllib3/contrib/appengine.py [hunk 4]
```diff
@@ -236,13 +236,21 @@ def _urlfetch_response_to_http_response(self, urlfetch_resp, **response_kw):
...
-            body=BytesIO(urlfetch_resp.content),
-            headers=urlfetch_resp.headers,
-            status=urlfetch_resp.status_code,
-            **response_kw
+          body=BytesIO(urlfetch_resp.content),
+          headers=urlfetch_resp.headers,
+          status=urlfetch_resp.status_code,
+          original_response=original_response,
+          **response_kw
         )
```
Comment (Lukasa) [inline urllib3/contrib/appengine.py]: Same here.

File: urllib3/contrib/appengine.py (hunk 3 above)
Comment (Lukasa) [reply urllib3/contrib/appengine.py]: Also can you match the house style of having the closing parenthesis on its own line?

**Comment by erlichmen** [comment]
@Lukasa Intent fixed.

File: urllib3/response.py (hunk 1 above)
Comment (sigmavirus24) [reply urllib3/response.py]: Sorry, I guess my question was not clear enough. Let me rephrase: Why can't we more simply do `self.msg = msg`?

**Comment by erlichmen** [comment]
This is a read only property, I don’t see a reason to allow modifying it
after the ctor

**Comment by sigmavirus24** [comment]
@erlichmen given the usecase it's being added to support, we don't need that boilerplate, in other words, since we don't anticipate people modifying it, why are we even bothering to go through with this. Let's keep things simple and if we find there's a convincing case for this boilerplate, we can change it.

**Comment by erlichmen** [comment]
@sigmavirus24 done

File: urllib3/response.py [hunk 5]
```diff
@@ -572,7 +576,6 @@ def read_chunked(self, amt=None, decode_content=None):
         """
         Similar to :meth:`HTTPResponse.read`, but with an additional
         parameter: ``decode_content``.
-
```
Comment (sigmavirus24) [inline urllib3/response.py]: Please add this new-line back?

**Comment by erlichmen** [comment]
@sigmavirus24 done (+ PR is now squashed)
@Lukasa take another look

**Comment by sigmavirus24** [comment]
That all looks good. Now we can focus on test coverage. I think most of what we're missing is your new method for `isclosed`. Adding some tests around that should get things to pass.

**Comment by trickidicki** [comment]
Hoping this can get resolved quickly as it's a blocking issue for me at the moment...

**Comment by sigmavirus24** [comment]
@trickidicki while I appreciate that this is preventing you from completing whatever it is you're working on, this patch will not be resolved until the code coverage is fixed. It seems like that hasn't happened since my last comment. If you want it to move along, perhaps you should write some tests for this to make it pass coverage. Otherwise, please be patient. @erlichmen, @haikuginger and I all have personal and professional lives that come before this project and your deadlines.

**Comment by erlichmen** [comment]
I add to an existing test assert over the new code, hopefully that eventually it will show up here.

**Comment by sigmavirus24** [comment]
Closing and re-opening to give AppVeyor a kick

File: urllib3/contrib/appengine.py [hunk 6]
```diff
@@ -236,12 +236,21 @@ def _urlfetch_response_to_http_response(self, urlfetch_resp, **response_kw):
...
             # In order for decoding to work, we must present the content as
             # a file-like object.
+            body=BytesIO(urlfetch_resp.content),
+            msg=urlfetch_resp.header_msg,
+            headers=urlfetch_resp.headers,
+            status=urlfetch_resp.status_code,
+            **response_kw
+        )
+
+        return HTTPResponse(
```
Comment (haikuginger) [inline urllib3/contrib/appengine.py]: Minor question: why are we setting `msg` on `original_response`, but not on the response object we return?

**Comment by erlichmen** [comment]
Only the code flow that needs original_response is using the msg field. I
don’t want to create a “new” API just because the internal code needs it

**Comment by haikuginger** [comment]
@erlichmen, great, thanks! Can you rebase this on master? It seems to be having some issues with our coverage system.

**Comment by erlichmen** [comment]
@haikuginger done
